mixer/adapter/rbac/rbac.go

```diff
@@ -0,0 +1,176 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:generate $GOPATH/src/istio.io/istio/bin/mixer_codegen.sh -f mixer/adapter/rbac/config/config.proto
+
+package rbac
+
+import (
+	"context"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+
+	rbacproto "istio.io/api/rbac/v1alpha1"
+	"istio.io/istio/mixer/adapter/rbac/config"
+	"istio.io/istio/mixer/pkg/adapter"
+	mixerconfig "istio.io/istio/mixer/pkg/config"
+	"istio.io/istio/mixer/pkg/config/store"
+	"istio.io/istio/mixer/pkg/status"
+	"istio.io/istio/mixer/template/authorization"
+)
+
+type (
+	builder struct {
+		adapterConfig *config.Params
+	}
+	handler struct {
+		rbac Rbac
+		env  adapter.Env
+	}
+)
+
+// ServiceRoleKind defines the config kind name of ServiceRole.
+const serviceRoleKind = "ServiceRole"
+
+// ServiceRoleBindingKind defines the config kind name of ServiceRoleBinding.
+const serviceRoleBindingKind = "ServiceRoleBinding"
+
+///////////////// Configuration-time Methods ///////////////
+
+func (b *builder) SetAdapterConfig(cfg adapter.Config) {
+	b.adapterConfig = cfg.(*config.Params)
+}
+
+func (b *builder) Validate() (ce *adapter.ConfigErrors) {
+	return
+}
+
+func (b *builder) SetAuthorizationTypes(types map[string]*authorization.Type) {}
+
+func (b *builder) Build(ctx context.Context, env adapter.Env) (adapter.Handler, error) {
+	reg := store.NewRegistry(mixerconfig.StoreInventory()...)
+	s, err := reg.NewStore(b.adapterConfig.ConfigStoreUrl)
+	if err != nil {
+		env.Logger().Errorf("unable to connect to the configuration server: %v", err)
+		return nil, err
+	}
+	r := &configStore{}
+	err = startController(s, r, env)
+	if err != nil {
+		env.Logger().Errorf("unable to start controller: %v", err)
+		return nil, err
+	}
+
+	return &handler{rbac: r, env: env}, nil
+}
+
+// maxEvents is the likely maximum number of events
+// we can expect in a second. It is used to avoid slice reallocation.
+const maxEvents = 50
+
+var watchFlushDuration = time.Second
+
+// startController creates a controller from the given params.
+func startController(s store.Store, r *configStore, env adapter.Env) error {
+	data, watchChan, err := startWatch(s)
+	if err != nil {
+		env.Logger().Errorf("Error while starting watching CRDs: %v", err)
+		return err
+	}
+
+	c := &controller{
+		configState: data,
+		rbacStore:   r,
+	}
+
+	c.processRbacRoles(env)
+	// Start a goroutine to watch for changes on a channel and
+	// publishes a batch of changes via applyEvents.
+	env.ScheduleDaemon(func() {
+		// consume changes and apply them to data indefinitely
+		var timeChan <-chan time.Time
+		var timer *time.Timer
+		events := make([]*store.Event, 0, maxEvents)
+
+		for {
+			select {
+			case ev := <-watchChan:
+				if len(events) == 0 {
+					timer = time.NewTimer(watchFlushDuration)
```

Comment:
i wonder... is it better to `Reset()` the existing Timer than to create a new one?  I don't actually know.